core/trino-main/src/main/java/io/trino/operator/output/IntPositionsAppender.java

```diff
@@ -14,31 +14,229 @@
 package io.trino.operator.output;
 
 import io.trino.spi.block.Block;
-import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.BlockBuilderStatus;
+import io.trino.spi.block.BlockUtil;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.IntArrayBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.trino.spi.block.BlockUtil.calculateBlockResetSize;
+import static java.lang.Math.max;
 
 public class IntPositionsAppender
-        implements PositionsAppender
+        implements BlockTypeAwarePositionsAppender
 {
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(IntPositionsAppender.class).instanceSize();
+    private static final Block NULL_VALUE_BLOCK = new IntArrayBlock(1, Optional.of(new boolean[] {true}), new int[1]);
+
+    @Nullable
+    private final BlockBuilderStatus blockBuilderStatus;
+    private boolean initialized;
+    private final int initialEntryCount;
+
+    private int positionCount;
+    private boolean hasNullValue;
+    private boolean hasNonNullValue;
+
+    // it is assumed that these arrays are the same length
+    private boolean[] valueIsNull = new boolean[0];
+    private int[] values = new int[0];
+
+    private long retainedSizeInBytes;
+
+    public IntPositionsAppender(@Nullable BlockBuilderStatus blockBuilderStatus, int expectedEntries)
+    {
+        this.blockBuilderStatus = blockBuilderStatus;
+        this.initialEntryCount = max(expectedEntries, 1);
+
+        updateDataSize();
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block block)
+    {
+        int[] positionArray = positions.elements();
+        int newPositionCount = positions.size();
+        ensureCapacity(positionCount + newPositionCount);
+
+        if (block.mayHaveNull()) {
+            for (int i = 0; i < newPositionCount; i++) {
+                int position = positionArray[i];
+                boolean isNull = block.isNull(position);
```

Comment:
you can extract:
```
                boolean isNull = block.isNull(position);
                int positionIndex = positionCount + i;
                if (isNull) {
                    valueIsNull[positionIndex] = true;
                    hasNullValue = true;
                }
                else {
                    values[positionIndex] = block.getInt(position, 0);
                    hasNonNullValue = true;
                }
```
and reuse it in `appendDictionary`

similar for non null version of the method

Reply:
Why would there be a pollution. `block` in this case should be a primitive block, no?